### ISSUE TITLE
Communicate download progress to Census API users

### ISSUE THREAD

**Issue by hthomas-czi** [chanzuckerberg/cellxgene-census#1033]
[Visibility of system status](https://www.nngroup.com/articles/ten-usability-heuristics/#1-visibility-of-system-status) is an important usability heuristic. Users want to know how long a data fetching call will take and how much progress is being made.

<Start Pablo's edit>
Applies to

`SOMA.DataFrame.read().concat()`
`SOMA.SparseNDArray.read().concat()`
`cellxgene_census.download_source_h5ad()`
<End Pablo's edit>

**User Quote**
> I’d like to see progress bars to understand how long to wait for download to finish

**Comment by pablo-gar** [comment]
@ebezzi for the SOMA-related task let's ask TileDB if there is any possibility to implement anything on this.

**Comment by ivirshup** [comment]
This is very straightforward for `cellxgene_census.download_source_h5ad`. Basically just:

```python
from fsspec.callbacks import TqdmCallback

...

    fs = s3fs.S3FileSystem(
        anon=True,
        cache_regions=True,
    )
    fs.get_file(
        locator["uri"],
        to_path,
        callback=TqdmCallback(),
    )
```

![image](https://github.com/chanzuckerberg/cellxgene-census/assets/8238804/2acf241b-79fb-4d32-aece-a9556e809354)

It sounds like using this for other things is more of a SOMA issue. 

* Should it be possible to turn this reporting off?
* Any ideas on how to test this?

**Comment by prathapsridharan** [comment]
In regards to progress for other APIs - Is supplying way of getting progress for programmatic APIs a common expectation? 

`download_*` seems straightforward enough but I can't recall expecting a progress indicator when programmatically invoking APIs on a terminal or script.

Would be good to have examples from other domains where programmatic usage of an API also has a mechanism to communicate progress of work.

**Comment by ivirshup** [comment]
> download_* seems straightforward enough but I can't recall expecting a progress indicator when programmatically invoking APIs on a terminal or script.

I think giving the option to report progress (sometimes by default) is pretty common, `wget` for instance. `dask` also has a number of tools for reporting progress of ongoing tasks. A lot of bioinformatics tools (though especially older ones) also make this available.

It's nice when its done via stderr or some alternative stream.

**Comment by pablo-gar** [comment]
I think something that can suffice the request and is simple for now for `SOMA.DataFrame.read().concat()` and `SOMA.SparseNDArray.read().concat()` is:

Adding a report progress (either via progress bar, or stderr stdout dumps) in each iteration of the [`_arrow_table_reader`](https://github.com/single-cell-data/TileDB-SOMA/blob/d837dd251323abcff6d81c187555c56b5431515c/apis/python/src/tiledbsoma/_read_iters.py#L460). 

@ivirshup what do you think? We can create a mini proposal for that change and discuss with TileDB folks.

**Comment by ivirshup** [comment]
> Adding a report progress (either via progress bar, or stderr stdout dumps) in each iteration of the [_arrow_table_reader](https://github.com/single-cell-data/TileDB-SOMA/blob/d837dd251323abcff6d81c187555c56b5431515c/apis/python/src/tiledbsoma/_read_iters.py#L460).

I think that would work. 

I also would really like to get some dask integration in tiledb-soma, which I think also give this ability.

**Comment by ivirshup** [comment]
@pablo-gar are you happy with the solution for `download_source_h5ad` in #1153?

To me, the main question left there is: do we want a way to turn this off?

**Comment by pablo-gar** [comment]
Yes we should give the user the ability to turn it off.

**Comment by ivirshup** [comment]
Cool.

I'll add a `progress_bar: bool = True` argument to the function?

**Comment by pablo-gar** [comment]
That works!

**Comment by ivirshup** [comment]
The PR is now ready for review. I also added units to the progress bar.

**Comment by ivirshup** [comment]
Fixed by #1153
